x-pack/plugins/ml/server/models/calculate_model_memory_limit/calculate_model_memory_limit.ts

```diff
@@ -0,0 +1,198 @@
+ */
+async function getCardinalities(
+  callAsCurrentUser: APICaller,
+  analysisConfig: AnalysisConfig,
+  indexPattern: string,
+  query: any,
+  timeFieldName: string,
+  earliestMs: number,
+  latestMs: number
+): Promise<{
+  overallCardinality: { [key: string]: number };
+  maxBucketCardinality: { [key: string]: number };
+}> {
+  /**
+   * Fields not involved in cardinality check
+   */
+  const excludedKeywords = new Set<string>(
+    /**
+     * The keyword which is used to mean the output of categorization,
+     * so it will have cardinality zero in the actual input data.
+     */
+    'mlcategory'
+  );
+
+  const fieldsService = fieldsServiceProvider(callAsCurrentUser);
+
+  const { detectors, influencers, bucket_span: bucketSpan } = analysisConfig;
+
+  let overallCardinality = {};
+  let maxBucketCardinality = {};
+  try {
+    const overallCardinalityFields: Set<string> = detectors.reduce(
+      (
+        acc,
+        {
+          by_field_name: byFieldName,
+          partition_field_name: partitionFieldName,
+          over_field_name: overFieldName,
+        }
+      ) => {
+        [byFieldName, partitionFieldName, overFieldName]
+          .filter(field => field !== undefined && field !== '' && !excludedKeywords.has(field))
+          .forEach(key => {
+            acc.add(key as string);
+          });
+        return acc;
+      },
+      new Set<string>()
+    );
+
+    const maxBucketFieldCardinalities: string[] = influencers.filter(
+      influencerField =>
+        typeof influencerField === 'string' &&
+        !excludedKeywords.has(influencerField) &&
+        !!influencerField &&
+        !overallCardinalityFields.has(influencerField)
+    ) as string[];
+
+    if (overallCardinalityFields.size > 0) {
+      overallCardinality = await fieldsService.getCardinalityOfFields(
+        indexPattern,
+        [...overallCardinalityFields],
+        query,
+        timeFieldName,
+        earliestMs,
+        latestMs
+      );
+    }
+
+    if (maxBucketFieldCardinalities.length > 0) {
+      maxBucketCardinality = await fieldsService.getMaxBucketCardinalities(
+        indexPattern,
+        maxBucketFieldCardinalities,
+        query,
+        timeFieldName,
+        earliestMs,
+        latestMs,
+        bucketSpan
+      );
+    }
+
+    return {
+      overallCardinality,
+      maxBucketCardinality,
+    };
+  } catch (e) {
+    throw new Error('Unable to retrieve cardinality of the partition fields or the influencers');
+  }
+}
+
+export function calculateModelMemoryLimitProvider(callAsCurrentUser: APICaller) {
+  /**
+   * Retrieves an estimated size of the model memory limit used in the job config
+   * based on the cardinality of the fields being used to split the data
+   * and influencers.
+   */
+  return async function calculateModelMemoryLimit(
+    analysisConfig: AnalysisConfig,
+    indexPattern: string,
+    query: any,
+    timeFieldName: string,
+    earliestMs: number,
```

Comment:
Just a note that in the follow-up bit of work, where the modules `setup` endpoint calls this `calculateModelMemoryLimit` function, that we need to handle the case where earliest and latest times are not supplied. In this case, the `setup` endpoint should find the start/end times of the latest 3 months of data, and pass those times to this function.

Reply:
Thanks for the heads-up, yes, I planned this changes for the follow-up PR dedicated to the setup endpoint